packages/@lwc/ssr-runtime/src/lightning-element.ts

```diff
@@ -70,9 +70,15 @@ export class LightningElement implements PropsAvailableAtConstruction {
         assign(this, propsAvailableAtConstruction);
     }
 
-    [SYMBOL__SET_INTERNALS](props: Properties, attrs: Attributes) {
+    [SYMBOL__SET_INTERNALS](props: Properties, attrs: Attributes, publicProperties: Array<string>) {
         this.#attrs = attrs;
-        assign(this, props);
+
+        // Set the property if it is part of the component's public API
+        for (const [propName, propValue] of Object.entries(props)) {
+            if (publicProperties.includes(propName)) {
```

Comment:
This should be a `Set` rather than an `Array` so we can do a fast `has` lookup rather than a slow `includes` lookup. I can imagine components with lots of `@api` props.

Reply:
Originally, I didn't think that optimization was necessary because a component that has enough public props to make this slow is probably an edge case and this is only done as an initialization step, but it's a small change so I added it!

Reply:
It is definitely debatable since the `Set` is overkill for ~2-3 items.